src/Ace/Halogen/Component.purs

```diff
@@ -1,75 +1,213 @@
+foreign import dataset
+  :: forall e. HTMLElement -> Eff (dom :: DOM |e) (StrMap String)
+
+
+-- | Take completion function for currently selected component
+completeFnFocused :: forall e. Eff (AceEffects e) (CompleteFn e)
+completeFnFocused = do
+  focusedKey <- readRef focused
+  mFns <- readRef completeFns
+  maybe (pure emptyCompleteFn) pure $ Sm.lookup focusedKey mFns
+  where
+  emptyCompleteFn :: CompleteFn e
+  emptyCompleteFn _ _ _ _ = pure []
+
+-- | Set autocomplete resume
+setAutocompleteResume
+  :: forall e. Maybe Autocomplete -> Editor -> Eff (AceEffects e) Unit
+setAutocompleteResume Nothing editor = do
+  Editor.setEnableBasicAutocompletion false editor
+  Editor.setEnableLiveAutocompletion false editor
+setAutocompleteResume (Just Basic) editor = do
+  Editor.setEnableLiveAutocompletion false editor
+  Editor.setEnableBasicAutocompletion true editor
+setAutocompleteResume (Just Live) editor = do
+  Editor.setEnableLiveAutocompletion true editor
+  Editor.setEnableBasicAutocompletion true editor
+
+-- | Language tools and autocomplete initializer. Runs once.
+enableAutocomplete :: forall e. Eff (AceEffects e) Unit
+enableAutocomplete = do
+  languageToolsInitialized <- readRef initialized
+  when (not languageToolsInitialized) do
+    completer <- Completer.mkCompleter globalCompleteFn
+    tools <- LanguageTools.languageTools
+    LanguageTools.addCompleter completer tools
+    writeRef initialized true
+  where
+  globalCompleteFn editor session position prefix cb = do
+    fn <- completeFnFocused
+    runAff (const $ cb Nothing) (cb <<< Just)
+      $ fn editor session position prefix
+
+-- | Generate unique key for component
+genKey :: forall e. Eff (now :: Now, random :: RANDOM|e) String
+genKey = do
+  rn1 <- random
+  rn2 <- random
+  (Milliseconds time) <- nowEpochMilliseconds
+  pure $ show rn1 <> show time <> show rn2
+
+----------------------------------------------------------------------
+-- Types, State and other common halogen things
+----------------------------------------------------------------------
+data Autocomplete = Live | Basic
+
+type AceEffects e =
+  ( random :: RANDOM
+  , now :: Now
+  , ref :: REF
+  , ace :: ACE
+  , avar :: AVAR
+  , dom :: DOM
+  | e)
+
+
+-- | Ace query algebra
+-- | `Init` used internally to handle initialization of component
+-- | `GetText` gets the current text value
+-- | `SetText` alters the current text value
+-- | `SetAutocomplete` sets autocomplete resume:
+-- |    `Nothing` turns it off
+-- |    `Just Basic` enables basic autocompletions (triggered by `Alt+Space` or `Ctrl + Space`)
+-- |    `Just Live` enables live autocomplete
+-- | `SetCompleteFn` sets function providing autocomplete variants.
+-- | `GetEditor` returns ace editor instance handled by this component.
+data AceQuery a
+  = Init HTMLElement a
+  | GetText (String -> a)
+  | SetText String a
+  | SetAutocomplete (Maybe Autocomplete) a
+  | SetCompleteFn (forall e. CompleteFn e) a
```

Comment:
Because live autocomplete can be really annoying -- it tries to complete every word. And with separate option user can add checkbox, keycombo or something else to turn it off or switch to basic.